pkg/controllers/cassandra/pilot/pilot.go

```diff
@@ -0,0 +1,165 @@
+	"k8s.io/apimachinery/pkg/labels"
+	appslisters "k8s.io/client-go/listers/apps/v1beta1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+type Interface interface {
+	Sync(*v1alpha1.CassandraCluster) error
+}
+
+type pilotControl struct {
+	naviClient   navigator.Interface
+	pilots       navlisters.PilotLister
+	pods         corelisters.PodLister
+	statefulSets appslisters.StatefulSetLister
+	recorder     record.EventRecorder
+}
+
+var _ Interface = &pilotControl{}
+
+func NewControl(
+	naviClient navigator.Interface,
+	pilots navlisters.PilotLister,
+	pods corelisters.PodLister,
+	statefulSets appslisters.StatefulSetLister,
+	recorder record.EventRecorder,
+) *pilotControl {
+	return &pilotControl{
+		naviClient:   naviClient,
+		pilots:       pilots,
+		pods:         pods,
+		statefulSets: statefulSets,
+		recorder:     recorder,
+	}
+
+}
+
+func (c *pilotControl) clusterPods(cluster *v1alpha1.CassandraCluster) ([]*v1.Pod, error) {
+	var clusterPods []*v1.Pod
+	allPods, err := c.pods.Pods(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return clusterPods, err
+	}
+	for _, pod := range allPods {
+		podControlledByCluster, err := controllers.PodControlledByCluster(
+			cluster,
+			pod,
+			c.statefulSets,
+		)
+		if err != nil {
+			return clusterPods, err
+		}
+		if !podControlledByCluster {
+			continue
+		}
+		clusterPods = append(clusterPods, pod)
+	}
+	return clusterPods, nil
+}
+
+func (c *pilotControl) createOrUpdatePilot(cluster *v1alpha1.CassandraCluster, pod *v1.Pod) error {
+	desiredPilot := PilotForCluster(cluster, pod)
+	client := c.naviClient.NavigatorV1alpha1().Pilots(desiredPilot.GetNamespace())
+	lister := c.pilots.Pilots(desiredPilot.GetNamespace())
+	existingPilot, err := lister.Get(desiredPilot.GetName())
+	if k8sErrors.IsNotFound(err) {
+		_, err = client.Create(desiredPilot)
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	err = util.OwnerCheck(existingPilot, cluster)
+	if err != nil {
+		return err
+	}
+	desiredPilot = existingPilot.DeepCopy()
+	updatePilotForCluster(cluster, pod, desiredPilot)
+	_, err = client.Update(desiredPilot)
+	return err
+}
+
+func (c *pilotControl) removeUnusedPilots(
+	cluster *v1alpha1.CassandraCluster,
+) error {
+	expectedPilotNames := map[string]bool{}
+	clusterPods, err := c.clusterPods(cluster)
+	if err != nil {
+		return err
+	}
+	for _, pod := range clusterPods {
+		expectedPilotNames[pod.Name] = true
+	}
+	existingPilots, err := c.pilots.Pilots(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	client := c.naviClient.NavigatorV1alpha1().Pilots(cluster.Namespace)
+	for _, pilot := range existingPilots {
+		err := util.OwnerCheck(pilot, cluster)
+		if err != nil {
+			return err
+		}
+		_, found := expectedPilotNames[pilot.Name]
+		if !found {
+			err := client.Delete(pilot.Name, nil)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
```

Comment:
So I'm not too sure if we should be doing this right now (deleting 'unused' pilot resources). It's difficult to define unused, and in fact, the lack of a pod does not immediately imply that the Pilot resource is no longer in use. I'd rather do something along the lines of waiting N seconds since the last heartbeat (provided no corresponding pod exists) before deleting.

Right now the Elasticsearch controller doesn't actually clean up old Pilots, it just cleans up old StatefulSets (i.e. ones no longer specified as a nodepool on the ElasticsearchCluster resource). Should we make these the same? Consistency in this sort of thing probably makes it easier to reason about. At some point I think we can refactor the Pilot control loop to make it mostly generic between DB types.

Reply:
> So I'm not too sure if we should be doing this right now (deleting 'unused' pilot resources). It's difficult to define unused, and in fact, the lack of a pod does not immediately imply that the Pilot resource is no longer in use. I'd rather do something along the lines of waiting N seconds since the last heartbeat (provided no corresponding pod exists) before deleting.

But what's the bug that the heartbeat / delay would fix?
If there is a bug, then I can add the heartbeat / delay in a followup branch.
But for now, removing unused Pilot resources seems like good housekeeping.

Reply:
The heartbeat is a suggestion for how it can be handled, but as you say, should be in a follow up.

I think that we can't say the lack of pod = the pilot doesn't exist. In the ES controller for example, the 'master' Pilot uses Pilot resources to determine whether a cluster is in a state ready to scale down for example. This scenario, as far as I can tell, will cause failures if we auto-delete Pilots if no Pod exists:

* User requests scale down of cluster
* Navigator updates a Pilot resource to 'decommission' it
* Because the user is a sadist, or because *something* happened, the pod that is being decommissioned gets deleted.
* This *should* cause the StatefulSet controller to re-create the pod, really, as the Pilot is *not* finished decommissioning so it needs to come back in order to finish
* However, because we immediately deleted the Pilot resource upon the Pod disappearing, the 'knowledge' of those remaining documents/indices left on that node (i.e. the `pilot.status.elasticsearch.documentCount` field) has been lost.
* So now our master Pilot is not aware of those extra documents, nor the fact that a Pilot is being decommissioned either (as that knowledge has also been lost).
* At this point we've reached a bit of a weird state - especially if we consider that the *only* state about the state of the world is stored in the k8s api and anything in-memory should not be relied upon

I do agree we need to do house keeping, but I'm thinking we should consider *how* this is done carefully as 'unused' is a tricky term. Additionally, given the Pilot resource in the API is one of the users primary points for debugging issues with their clusters, by deleting the Pilots so frequently we are also destroying a lot of useful debugging information for users. I'd much prefer the behaviour between all DB types to be consistent - the first steps in debugging the ES cluster should be the same as Cassandra, and vice versa.

Reply:
^ my only outstanding/current issue with this PR before I'm happy to merge 😄 .

Let me know if you disagree, but any change we make here needs to be consistent with ES too given this touches a user-facing resource (the Pilot resource)

Reply:
It seems to me that in the scenario above,

* Accidentally deleted pod gets restarted by StatefulSet Controller.
* Pilot process waits for appearance of Pilot
* Pilot starts ES sub-process
* Pilot re-updates the Pilot status with document count 0
* navigator controller waits for document count 0 and then decrements the Statefulset replicas count 
* and removes the drained pod and corresponding pilot.

But I agree that we should keep ES and Cass in sync, so I'll remove the Pilot cleanup code for now.